arduino/cores/packagemanager/profiles.go

```diff
@@ -0,0 +1,188 @@
+// If installMissing is true then possibly missing tools and platforms will be downloaded and installed.
+func (pm *PackageManager) LoadHardwareForProfile(p *sketch.Profile, installMissing bool, downloadCB rpc.DownloadProgressCB, taskCB rpc.TaskProgressCB) []error {
+	// Load required platforms
+	var merr []error
+	var platformReleases []*cores.PlatformRelease
+	indexURLs := map[string]*url.URL{}
+	for _, platformRef := range p.Platforms {
+		if platformRelease, err := pm.loadProfilePlatform(platformRef, installMissing, downloadCB, taskCB); err != nil {
+			merr = append(merr, fmt.Errorf("%s: %w", tr("loading required platform %s", platformRef), err))
+			logrus.WithField("platform", platformRef).WithError(err).Debugf("Error loading platform for profile")
+		} else {
+			platformReleases = append(platformReleases, platformRelease)
+			indexURLs[platformRelease.Platform.Name] = platformRef.PlatformIndexURL
+			logrus.WithField("platform", platformRef).Debugf("Loaded platform for profile")
+		}
+	}
+
+	// Load tools dependencies for the platforms
+	for _, platformRelease := range platformReleases {
+		// TODO: pm.FindPlatformReleaseDependencies(platformRelease)
+
+		for _, toolDep := range platformRelease.ToolDependencies {
+			indexURL := indexURLs[toolDep.ToolPackager]
+			if err := pm.loadProfileTool(toolDep, indexURL, installMissing, downloadCB, taskCB); err != nil {
+				merr = append(merr, fmt.Errorf("%s: %w", tr("loading required tool %s", toolDep), err))
+				logrus.WithField("tool", toolDep).WithField("index_url", indexURL).WithError(err).Debugf("Error loading tool for profile")
+			} else {
+				logrus.WithField("tool", toolDep).WithField("index_url", indexURL).Debugf("Loaded tool for profile")
+			}
+		}
+	}
+
+	return merr
+}
+
+func (pm *PackageManager) loadProfilePlatform(platformRef *sketch.ProfilePlatformReference, installMissing bool, downloadCB rpc.DownloadProgressCB, taskCB rpc.TaskProgressCB) (*cores.PlatformRelease, error) {
+	targetPackage := pm.Packages.GetOrCreatePackage(platformRef.Packager)
+	platform := targetPackage.GetOrCreatePlatform(platformRef.Architecture)
+	release := platform.GetOrCreateRelease(platformRef.Version)
+
+	uid := platformRef.InternalUniqueIdentifier()
+	destDir := configuration.ProfilesCacheDir(configuration.Settings).Join(uid)
+	if !destDir.IsDir() && installMissing {
+		// Try installing the missing platform
+		if err := pm.installMissingProfilePlatform(platformRef, destDir, downloadCB, taskCB); err != nil {
+			return nil, err
+		}
+	}
+	return release, pm.loadPlatformRelease(release, destDir)
+}
+
+func (pm *PackageManager) installMissingProfilePlatform(platformRef *sketch.ProfilePlatformReference, destDir *paths.Path, downloadCB rpc.DownloadProgressCB, taskCB rpc.TaskProgressCB) error {
+	// TODO: TaskProgressCB
+
+	// Instantiate a temporary package manager only for platform installation
+	tmp, err := paths.MkTempDir("", "")
```

Comment:
If `directories.data` is on a different drive than the system temporary directory, this will result in the violation of the requirement stated here:
https://github.com/arduino/arduino-cli/blob/7dd09e72245c06f575ca1c7f53e77a5f7ae5771f/arduino/resources/install.go#L32-L33

example:

```
$ ../arduino-cli compile --profile foo
Downloading platform arduino:avr@1.8.5...
Downloading index: package_index.json downloaded
arduino:avr@1.8.5 already downloaded
Installing platform arduino:avr@1.8.5...
installing platform arduino:avr@1.8.5...
Error initializing instance: Error loading hardware platform: loading required platform arduino:avr@1.8.5: Error installing platform arduino:avr@1.8.5: installing platform arduino:avr@1.8.5: moving extracted archive to destination dir: rename C:\Users\per\AppData\Local\Temp\2481469040\package-1179567273\ArduinoCore-avr-1.8.5 e:\deleteme\arduino-cli\directories\data\internal\arduino_avr_1.8.5_70334d3546efee53: The system cannot move the file to a different disk drive.

Error during build: Platform 'arduino:avr' not found: platform not installed
Try running `arduino-cli.exe core install arduino:avr`
```

This situation is avoided by `core install` by placing the temporary directory under `directories.data`:

https://github.com/arduino/arduino-cli/blob/7dd09e72245c06f575ca1c7f53e77a5f7ae5771f/commands/instances.go#L139

Reply:
Nice catch!